_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+
+### Introducing User-Defined Operators
+The new feature extends the third form of the `using for` syntax to allow binding operators to functions.
+The function is invoked whenever the operator is used on the values of the type:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+using {add as +} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b; // Equivalent to add(a, b)
+}
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
+The type specified in `using for` must be a user-defined value type and must be the type
+of all the parameters of the function and its return value.
+The only exception are comparison operators, where the return type must be `bool`.
+
+Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
+must always be invoked with exact types.
+
+The following operators can be defined:
+| Category   | Operators                        |
+|------------|----------------------------------|
+| Bitwise    | `&`, `\|`, `^`,  `~`             |
+| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.
+Note that Solidity does not have unary `+`.
+Unary and binary `-` are considered different operators and need separate definitions.
+
+It is not possible to redefine built-in operators.
+All valid cases are the ones where previously there was no built-in operator.
+As such, only one definition is allowed for each user-defined operator for a given type.
+As long as the definition stays the same, it is not an error to repeat the `using for` directive,
+just as it is not an error to attach the same function to a type twice.
+
+#### Operator Bindings vs Attached Functions
+User-defined operators are independent from attached functions.
+One can be done without the other, and both can be done at the same time, even in the same `using for` directive.
+
+The following example illustrates this:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+using {add as +} for Int global;
+using {sub as -, sub} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function sub(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int x, Int y) pure {
+    x + y;
+    x.add(y); // ERROR: Member "add" not found or not visible after argument-dependent lookup in Int.
+
+    x - y;
+    x.sub(y); // OK -- "sub" was also attached in "using for"
+}
+```
+
+In the above example, the function `add()` is used *only* as the operator `+`, and not callable as `x.add()`.
```

Comment:
```suggestion
In the above example, the function `add()` is used *only* as the operator `+`, and not callable as `x.add()` (but it can of course be called like `add(x, y)`).
```